# Canon Rule #40: Legacy Exception & Annotation Contract

**Status:** ‚úÖ Mandatory  
**Category:** Governance / Enforcement  
**Automation:** Partial (Lint + Validator)  
**CI Block:** Conditional  
**Version:** 1.0.0  
**Date:** 2025-01-02

---

## üéØ Objective

Allow **explicit, trackable, and temporary exceptions** to canonical rules without breaking governance, without hiding technical debt, and without enabling anarchy.

‚ö†Ô∏è **Exceptions are not ignored violations**  
‚úÖ **They are documented contracts with deadline, owner, and justification**

---

## üß± Fundamental Principle

**Nothing can violate the Canon without leaving a trace.**

If something:
- Uses forbidden token
- Uses arbitrary pixels
- Breaks provider rule
- Maintains unavoidable legacy

üëâ **It must be annotated.**

---

## üß© Canonical Annotation Format

The annotation must exist at the top of the file, before any code.

### üìå Mandatory Format
```rust
//! @canon-level: loose
//! @canon-exceptions: [#21, #24]
//! @canon-justification: Legacy component prior to token system
//! @canon-owner: ui-team
//! @canon-target-date: 2025-03-01
//! @canon-migration-status: planned
```

### üßæ Fields Contract

| Field | Required | Description |
|-------|----------|-------------|
| `@canon-level` | ‚úÖ | `strict`, `standard`, or `loose` |
| `@canon-exceptions` | ‚úÖ | List of broken rules |
| `@canon-justification` | ‚úÖ | Why this exists |
| `@canon-owner` | ‚úÖ | Team or responsible person |
| `@canon-target-date` | ‚ö†Ô∏è | Elimination deadline |
| `@canon-migration-status` | ‚ö†Ô∏è | `planned` \| `in-progress` \| `blocked` |

---

## üéöÔ∏è Canon Levels (Real Impact)

### `strict`

‚ùå **No exceptions allowed**  
- CI blocks
- Used for:
  - New components
  - Providers
  - Critical infrastructure

### `standard`

‚ö†Ô∏è **Exceptions allowed with annotation**  
- CI warns
- Project default

### `loose`

üü° **Legacy tolerated**  
- CI doesn't block
- Validator tracks debt

---

## üîç Linter Behavior (Canon Rule #40)

### 1Ô∏è‚É£ Without Annotation ‚Üí Error
```
‚ùå Canon #40 violation
File uses non-canonical tokens but has no @canon-exceptions annotation
```

‚û°Ô∏è **CI fails**

### 2Ô∏è‚É£ With Valid Annotation ‚Üí Allowed
```
‚ö†Ô∏è Canon Exception (#21, #24)
Owner: ui-team
Target date: 2025-03-01
```

‚û°Ô∏è **CI continues**  
‚û°Ô∏è **Violation becomes tracked exception**

### 3Ô∏è‚É£ Invalid Annotation ‚Üí Error

Examples:
- Non-existent rule (#99)
- Missing field
- Wrong format

‚û°Ô∏è **CI fails**

---

## ü§ñ Automatic Lint Rules

### Token Validator Logic
```rust
if violation_found {
    if has_canon_annotation(file) {
        mark_as_exception()
    } else {
        fail()
    }
}
```

### What Linter Validates Automatically

‚úîÔ∏è Annotation presence  
‚úîÔ∏è Correct syntax  
‚úîÔ∏è Rule exists (1‚Äì40)  
‚úîÔ∏è Valid `canon-level`  
‚úîÔ∏è File matches configured level  
‚úîÔ∏è No exception in forbidden provider  

---

## üßÆ Impact on Compliance Score

Rule #40 doesn't "zero out" compliance, it **classifies debt**.

### Example:

| Situation | Impact |
|-----------|--------|
| Violation without annotation | ‚ùå -5% |
| Annotated exception | ‚ö†Ô∏è 0% |
| Canonical | ‚úÖ +0% |

üëâ **Result:**  
Compliance = real quality + transparency

---

## üìä Reports

In reports (`canonrs report`):
```
Total files: 182
Canonical: 130
Annotated legacy: 52
Untracked violations: 0

Compliance: 100% (tracked)
```

**This is gold for enterprise.**

---

## üß† Why This Rule is Canon's Differentiator

### Without Rule #40:
- Design systems become dogma
- Teams bypass rules
- Lints get disabled

### With Rule #40:
- Technical debt stays visible
- Compliance becomes metric
- Evolution is gradual
- CI isn't the enemy

üëâ **This rule transforms Canon from "tool" to healthy political system.**

---

## üìù Exception Categories

### 1. Legacy Migration

**When:** Pre-canon code being gradually upgraded
```rust
//! @canon-level: loose
//! @canon-exceptions: [#21, #24]
//! @canon-justification: Component predates token system
//! @canon-migration-status: planned
//! @canon-target-date: 2025-Q2
```

**Allowed violations:** Any (temporary)  
**Requires:** Migration plan + target date  
**Review frequency:** Monthly

---

### 2. Third-Party Integration

**When:** External library constraints
```rust
//! @canon-level: standard
//! @canon-exceptions: [#24]
//! @canon-justification: Stripe widget requires exact 44px height
//! @canon-vendor: Stripe
//! @canon-vendor-docs: https://stripe.com/docs/...
```

**Allowed violations:** Size/color matching external requirements  
**Requires:** Vendor documentation reference  
**Review frequency:** On vendor update

---

### 3. Brand Requirements

**When:** White-label or brand-specific overrides
```rust
//! @canon-level: standard
//! @canon-exceptions: [#21]
//! @canon-justification: ACME Corp brand colors (contract requirement)
//! @canon-customer: ACME Corp
//! @canon-contract-expires: 2025-12-31
```

**Allowed violations:** Customer-specific colors  
**Requires:** Contract reference + expiration  
**Review frequency:** On contract renewal

---

### 4. Performance Critical

**When:** Optimization requires non-canonical approach
```rust
//! @canon-level: standard
//! @canon-exceptions: [#35]
//! @canon-justification: Inline calc for 60fps animation
//! @canon-benchmark: benchmarks/animation-perf.md
```

**Allowed violations:** Performance-justified only  
**Requires:** Benchmark data  
**Review frequency:** Quarterly

---

## üîß Implementation Example

### Before (Violation)
```rust
// switch.rs
pub fn Switch() -> impl IntoView {
    view! {
        <button class="border-gray-300 p-[2px]">
            // ‚ùå Canon #21, #24 violations
        </button>
    }
}
```

**Validator output:**
```
‚ùå Canon #21: border-gray-300
‚ùå Canon #24: p-[2px]
```

### After (Annotated Exception)
```rust
//! @canon-level: loose
//! @canon-exceptions: [#21, #24]
//! @canon-justification: Legacy switch component
//! @canon-owner: ui-team
//! @canon-target-date: 2025-03-01
//! @canon-migration-status: planned

pub fn Switch() -> impl IntoView {
    view! {
        <button class="border-gray-300 p-[2px]">
            // ‚ö†Ô∏è Tracked exception
        </button>
    }
}
```

**Validator output:**
```
‚ö†Ô∏è Exception tracked (#21, #24)
Owner: ui-team | Target: 2025-03-01
```

---

## üö® Forbidden Patterns

### ‚ùå Exception Without Justification
```rust
//! @canon-level: loose
//! @canon-exceptions: [#21]
// ‚ùå NO JUSTIFICATION - INVALID
```

**Linter error:**
```
‚ùå Canon #40: Missing @canon-justification
```

---

### ‚ùå Exception for Non-Existent Rule
```rust
//! @canon-level: loose
//! @canon-exceptions: [#99]  // ‚ùå Rule doesn't exist
```

**Linter error:**
```
‚ùå Canon #40: Rule #99 does not exist
```

---

### ‚ùå Exception in Forbidden Area
```rust
// ‚ùå FORBIDDEN: Exception in provider
// packages-rust/rs-design/src/providers/theme_provider.rs

//! @canon-level: loose
//! @canon-exceptions: [#32]
// ‚ùå Providers cannot have exceptions for localStorage
```

**Linter error:**
```
‚ùå Canon #37: Providers cannot use exceptions for critical rules
```

---

## üìà Monthly Exception Review

### Automated Report
```bash
canonrs exceptions --expired
```

**Output:**
```
‚ö†Ô∏è EXPIRED EXCEPTIONS:

ui/switch.rs
  Owner: ui-team
  Target: 2025-01-01 (EXPIRED)
  Exceptions: #21, #24

ui/separator.rs
  Owner: ui-team
  Target: 2025-01-15 (15 days overdue)
  Exceptions: #21

Total expired: 2
```

---

## üéØ Success Metrics

### Healthy Exception Management:

- ‚úÖ 90%+ of exceptions have target dates
- ‚úÖ <10% of codebase has exceptions
- ‚úÖ Average exception lifetime <6 months
- ‚úÖ Zero expired exceptions >3 months old

### Warning Signs:

- ‚ö†Ô∏è Exceptions growing faster than removals
- ‚ö†Ô∏è Same files excepted repeatedly
- ‚ö†Ô∏è Expired exceptions ignored
- ‚ö†Ô∏è Generic justifications ("legacy", "TODO")

---

## üß™ Testing

### Valid Annotation Test
```rust
#[test]
fn test_valid_exception_annotation() {
    let content = r#"
//! @canon-level: loose
//! @canon-exceptions: [#21]
//! @canon-justification: Legacy component
//! @canon-owner: ui-team
    "#;
    
    assert!(is_valid_annotation(content));
}
```

### Expired Exception Test
```rust
#[test]
fn test_expired_exception_warning() {
    let annotation = parse_annotation(file);
    let target = parse_date(&annotation.target_date);
    
    if target < now() {
        warn!("Exception expired: {}", file);
    }
}
```

---

## üèÅ Conclusion

### ‚úîÔ∏è Yes, it fits  
### ‚úîÔ∏è Yes, it's necessary  
### ‚úîÔ∏è You already implemented the technical base  
### ‚úîÔ∏è Only formal documentation was missing  

**Canon Rule #40 is the peace treaty between ideal and reality.**

---

## üìö References

- [Canon Rule #35: Token Usage Validation](./canon-rule-35-token-usage-validation.md)
- [Canon Rule #36: Component Compliance Levels](./canon-rule-36-component-compliance-levels.md)
- [Canon Rule #37: Provider Taxonomy & Boundaries](./canon-rule-37-provider-taxonomy-boundaries.md)

---

**Enforcement:** Automated via validators  
**Exception Limit:** Max 10% of codebase  
**Review Cycle:** Monthly  
**Auto-expire:** 6 months without renewal
