# Canon Rule #50: Provider Singleton & Runtime Separation

**Status:** Normative  
**Severity:** High  
**Applies to:** All Provider-based reactive systems

---

## The Principle

### Definitions

**Interaction Scope:** A subtree responsible for a continuous unit of interaction (drag operation, editor session, workflow step). Each scope maintains its own reactive runtime.


**Runtime Providers must be unique per interaction scope.**

Providers create reactive runtime (signals, state, lifecycle). Duplicating them creates invisible parallel contexts â†’ ghost bugs.

**Corollary:** Design systems NEVER create Providers. Only the Application layer controls Provider lifecycle.

---

## The Problem

### âŒ Wrong Pattern (Duplicated Provider)
```rust
// App layer
<App>
  <DragDropProvider>  // âœ… Provider 1
    <Dashboard>
      <DragDropProvider>  // ğŸš« Provider 2 (DUPLICATE!)
        ...
      </DragDropProvider>
    </Dashboard>
  </DragDropProvider>
</App>
```

**Why this breaks:**
- Two separate `DragContext` instances created
- `DragHandle` uses Context 1
- `DropZone` uses Context 2
- Events emitted to Context 1, callbacks listening on Context 2
- **Result:** Drag & drop silently fails (no errors, just doesn't work)

### âŒ Wrong Pattern (Provider in Design System)
```rust
// rs-design/ui/dashboard/dashboard.rs
#[component]
pub fn Dashboard() -> impl IntoView {
    view! {
        <DragDropProvider>  // ğŸš« FORBIDDEN
            <div>...</div>
        </DragDropProvider>
    }
}
```

**Why this breaks:**
- Design system component now controls runtime lifecycle
- Impossible to reuse `Dashboard` without drag & drop
- Violates separation of concerns
- Creates duplicate provider when used in app that already has one

---

## The Solution

### âœ… Correct Pattern (Singleton per Scope)
```rust
// App layer ONLY
#[component]
pub fn App() -> impl IntoView {
    view! {
        <Router>
            <DragDropProvider>  // âœ… ONE provider at app root
                <DragDropCallbacksProvider>  // âœ… Nested correctly
                    <Dashboard />  // âœ… NO provider here
                    <WorkflowEditor />  // âœ… NO provider here
                </DragDropCallbacksProvider>
            </DragDropProvider>
        </Router>
    }
}
```
```rust
// rs-design: Design system component (NO provider)
#[component]
pub fn Dashboard(
    widgets: Signal<Vec<WidgetDef>>,
    on_drop: Callback<DropEvent>,
) -> impl IntoView {
    // âœ… Consumes context (does NOT create it)
    let drag_ctx = use_drag_context();
    
    view! {
        <div>
            <DragHandle .../>  // âœ… Uses existing context
            <DropZone .../>    // âœ… Uses existing context
        </div>
    }
}
```

---

## Architecture Layers

### Layer Responsibilities

| Layer | Creates Providers? | Consumes Context? | Responsibility |
|-------|-------------------|-------------------|----------------|
| **App (frontend-leptos)** | âœ… YES | âœ… YES | Lifecycle, orchestration |
| **Design System (rs-design)** | âŒ NO | âœ… YES | UI primitives, emit events |
| **Domain/Core** | âŒ NO | âŒ NO | Business logic |

### Canonical Provider Hierarchy
```
App Root
â””â”€â”€ DragDropProvider (runtime: creates signals, state)
    â””â”€â”€ DragDropCallbacksProvider (interprets events â†’ commands)
        â””â”€â”€ LanguageProvider
            â””â”€â”€ ThemeProvider
                â””â”€â”€ UI Components (consume contexts)
```

**Critical Rule:** Providers that depend on runtime MUST be nested INSIDE the runtime provider.

**Why:** `DragDropCallbacksProvider` interprets events from `DragDropProvider`. If callbacks are outside, they have no runtime to listen to.

---

## Valid Scenarios

### âœ… Multiple Isolated Scopes
```rust
// Valid: Each scope has its own runtime
<App>
  <Route path="/dashboard">
    <DragDropProvider>  // âœ… Scope 1
      <Dashboard />
    </DragDropProvider>
  </Route>
  
  <Route path="/kanban">
    <DragDropProvider>  // âœ… Scope 2 (isolated)
      <KanbanBoard />
    </DragDropProvider>
  </Route>
</App>
```

**Rule:** Multiple providers are valid if they control **different interaction scopes**.

### âœ… Conditional Provider
```rust
// Valid: Feature flag
{move || {
    if feature_flags.drag_enabled.get() {
        view! {
            <DragDropProvider>
                <Dashboard />
            </DragDropProvider>
        }.into_any()
    } else {
        view! {
            <Dashboard />  // âœ… Works without drag
        }.into_any()
    }
}}
```

---

## Forbidden Patterns âŒ

### ğŸš« Provider in rs-design
```rust
// rs-design/ui/dashboard.rs
pub fn Dashboard() -> impl IntoView {
    view! {
        <DragDropProvider>  // ğŸš« FORBIDDEN
            ...
        </DragDropProvider>
    }
}
```

### ğŸš« Nested Provider in Same Scope
```rust
<DragDropProvider>
  <Tabs>
    <TabContent>
      <DragDropProvider>  // ğŸš« DUPLICATE
        ...
      </DragDropProvider>
    </TabContent>
  </Tabs>
</DragDropProvider>
```

### ğŸš« Callbacks Provider Outside Runtime
```rust
// WRONG ORDER
<DragDropCallbacksProvider>  // ğŸš« No runtime to listen to!
  <DragDropProvider>
    ...
  </DragDropProvider>
</DragDropCallbacksProvider>
```

### ğŸš« Hidden Provider in Reusable Component
```rust
// Some reusable component
pub fn ReusableWidget() -> impl IntoView {
    view! {
        <DragDropProvider>  // ğŸš« Hidden coupling
            <div>...</div>
        </DragDropProvider>
    }
}
```

---

## SSR & WASM Rules

### âœ… SSR-Safe Provider
```rust
#[component]
pub fn DragDropProvider(children: Children) -> impl IntoView {
    #[cfg(target_arch = "wasm32")]
    {
        let context = DragContext::new();  // âœ… Only in WASM
        provide_context(context);
    }
    
    children()
}
```

### ğŸš« Unsafe: Signal Creation in SSR
```rust
pub fn DragDropProvider() -> impl IntoView {
    // ğŸš« Creates RwSignal in SSR â†’ spawn_local panic
    let context = DragContext::new();
    provide_context(context);
}
```

**Rule:** Reactive signals MUST NOT be created during SSR unless guarded.

---

## Real-World Example: Dashboard Editor

### âŒ Before (Broken)
```rust
// App
<App>
  <DragDropProvider>
    <DashboardTab />
  </DragDropProvider>
</App>

// rs-design/dashboard.rs
pub fn Dashboard() -> impl IntoView {
    view! {
        <DragDropProvider>  // ğŸš« DUPLICATE
            ...
        </DragDropProvider>
    }
}
```

**Symptom:** Drag events fire, but widgets don't appear. No errors logged.

**Cause:** Two contexts. `DragHandle` emits to Context A, `DropZone` listens on Context B.

### âœ… After (Fixed)
```rust
// App (ONE provider at root)
<App>
  <DragDropProvider>
    <DragDropCallbacksProvider>
      <DashboardTab />
    </DragDropCallbacksProvider>
  </DragDropProvider>
</App>

// rs-design/dashboard.rs (NO provider)
pub fn Dashboard() -> impl IntoView {
    let ctx = use_drag_context();  // âœ… Uses existing
    view! { ... }
}

// frontend-leptos/dashboard_tab.rs
pub fn DashboardTab() -> impl IntoView {
    let callbacks = use_context::<DragDropCallbacks>()
        .expect("DragDropCallbacks required");
    
    let on_drop = Callback::new(|event| { ... });
    callbacks.register_drop(on_drop);  // âœ… Registers to shared context
}
```

**Result:** Single context. All components share same runtime. Drag & drop works.

---

## Benefits

### âœ… Predictable Context Resolution
- One provider â†’ one context
- No ambiguity about which context is active
- Leptos `use_context()` always finds the right one

### âœ… Reusable Components
```rust
// Dashboard works with OR without drag
<Dashboard widgets=.../>  // âœ… No drag (just displays)

<DragDropProvider>
  <Dashboard widgets=.../>  // âœ… With drag (interactive)
</DragDropProvider>
```

### âœ… Testable
```rust
#[test]
fn test_dashboard_no_drag() {
    // âœ… Can test without provider
    let dashboard = Dashboard::new(widgets);
    assert!(dashboard.renders());
}
```

### âœ… SSR-Safe
- Providers only create runtime in WASM
- No `spawn_local` panics
- Server can render static HTML

---

## Debug Checklist

When drag & drop doesn't work:

- [ ] **Check for duplicate providers**
```bash
  grep -r "DragDropProvider" src/
  # Should appear ONCE in app.rs
```

- [ ] **Verify provider order**
```
  Runtime BEFORE Callbacks
  DragDropProvider > DragDropCallbacksProvider
```

- [ ] **Check SSR guards**
```rust
  #[cfg(target_arch = "wasm32")]
```

- [ ] **Verify context consumption**
```rust
  // Should NOT panic
  let ctx = use_drag_context();
```

- [ ] **Check for hidden providers in components**
```bash
  grep -r "provide_context" rs-design/
  # Should be EMPTY
```

---

## Comparison: CanonRS vs Others

| Framework | Provider Pattern | Duplication Risk |
|-----------|------------------|------------------|
| **CanonRS** | Explicit singleton | âœ… Enforced by rule |
| React Context | Per-component | âš ï¸ Easy to duplicate |
| Vue Provide/Inject | Hierarchical | âš ï¸ Ambiguous scopes |
| Svelte Context | Module-level | âœ… Natural singleton |

**Veredito:** CanonRS is **more explicit** and **less error-prone** than React/Vue.

---

## Related Rules

- **Rule #49:** Drag & Drop as Intent (Event separation)
- **Rule #8:** Overlay Islands (Client-only architecture)
- **Rule #XX:** Separation of Concerns (pending)

---

## Testing Requirements

**E2E Tests:**
- E2E tests MUST mount the same providers as production
- Test environment MUST replicate provider hierarchy
- Mock providers MUST maintain same context structure


## Normative Requirements

**MUST:**
- Runtime providers MUST be unique per interaction scope
- Design system MUST NOT create providers
- Providers MUST be guarded for SSR when creating signals
- Provider order MUST respect dependencies (runtime before callbacks)

**MUST NOT:**
- Duplicate providers in same functional scope
- Hide providers inside reusable components
- Create reactive signals in SSR without guards

**SHOULD:**
- Place providers at app root or route level
- Document provider dependencies explicitly
- Test components with and without providers

---

**Author:** Canon Working Group  
**Date:** 2026-01-04  
**Version:** 1.0  
**Supersedes:** Implicit provider patterns  
**Related:** Canon Rule #49 (Drag & Drop as Intent)
